### PR TITLE
don't clear mkldnn cache in block_op executor dtor

### PR DESCRIPTION
**Pull request by sfraczek**
### PR types
Bug fixes

### PR changes
Others

### Describe
MKLDNN cache is removed in Executor's destructor. This should not happen at the end of RunImpl of conditional_block_op, where locally created Executor is destroyed. When working on dygraph resnet enablement of MKLDNN, I found that `test_resnet.py` with MKLDNN enabled will crash because it cannot find **forward MKLDNN primitive** in cache.

### CODE REVIEW

**Review by grygielski — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/distributed_ops/fl_listen_and_serv_op.cc [hunk 1]
```diff
@@ -218,6 +218,9 @@ void FlListenAndServOp::RunImpl(const framework::Scope &scope,
       "optimize blocks should be 1 at least on the pserver side.");
   auto *program = optimize_blocks[0]->Program();
   framework::Executor executor(dev_place);
+#ifdef PADDLE_WITH_MKLDNN
+  exec.KeepMKLDNNCache(true);
```
Comment (wojtuss) [inline paddle/fluid/operators/distributed_ops/fl_listen_and_serv_op.cc]: `executor`

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc [hunk 2]
```diff
@@ -99,6 +99,9 @@ class CGenNCCLIdOp : public framework::OperatorBase {
 
     framework::ProgramDesc empty_program;
     framework::Executor executor(dev_ctx.GetPlace());
+#ifdef PADDLE_WITH_MKLDNN
+    exec.KeepMKLDNNCache(true);
```
Comment (wojtuss) [inline paddle/fluid/operators/collective/c_gen_nccl_id_op.cc]: `executor`

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc [hunk 3]
```diff
@@ -214,6 +214,9 @@ class GenNCCLIdOp : public framework::OperatorBase {
 
     framework::ProgramDesc empty_program;
     framework::Executor executor(dev_ctx.GetPlace());
+#ifdef PADDLE_WITH_MKLDNN
+    exec.KeepMKLDNNCache(true);
```
Comment (wojtuss) [inline paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc]: `executor`

File: paddle/fluid/operators/distributed_ops/listen_and_serv_op.cc [hunk 4]
```diff
@@ -401,6 +401,9 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
                         "should be 1 at least on the pserver side."));
   auto *program = optimize_blocks[0]->Program();
   framework::Executor executor(dev_place);
+#ifdef PADDLE_WITH_MKLDNN
+  exec.KeepMKLDNNCache(true);
```
Comment (wojtuss) [inline paddle/fluid/operators/distributed_ops/listen_and_serv_op.cc]: `executor`

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 5]
```diff
@@ -134,6 +134,9 @@ class TensorRTEngineOp : public framework::OperatorBase {
   void RunNativeImpl(const framework::Scope &scope,
                      const platform::Place &dev_place) const {
     framework::Executor executor(dev_place);
+#ifdef PADDLE_WITH_MKLDNN
+    exec.KeepMKLDNNCache(true);
```
Comment (wojtuss) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h]: `executor`

**Comment by grygielski** [comment]
@luotao1 PR-CI-Coverage fails because of test coverage in `fl_listen_and_serve_op` and `conditional_block_infer_op`. However, these lines are not executed in tests, only because test build does not have `WITH_MKLDNN` flag. Could you advice how we can pass this CI?

**Comment by luotao1** [comment]
> When working on dygraph resnet enablement of MKLDNN, I found that test_resnet.py with MKLDNN enabled will crash because it cannot find forward MKLDNN primitive in cache.

Is this related with don't clear mkldnn cache at the end of RunImpl of conditional_block_op?

> MKLDNN cache is removed in Executor's destructor. This should not happen at the end of RunImpl of conditional_block_op, where locally created Executor is destroyed.

Do you have any other method to solve this problem? Since the current method is not grace.

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc [hunk 6]
```diff
@@ -99,6 +99,9 @@ class CGenNCCLIdOp : public framework::OperatorBase {
 
     framework::ProgramDesc empty_program;
     framework::Executor executor(dev_ctx.GetPlace());
+#ifdef PADDLE_WITH_MKLDNN
+    executor.KeepMKLDNNCache(true);
+#endif
```
Comment (luotao1) [inline paddle/fluid/operators/collective/c_gen_nccl_id_op.cc:104]: c_gen_nccl_id_op.cc is used only in GPU, thus, it doesn't need to be updated.

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc [hunk 7]
```diff
@@ -214,6 +214,9 @@ class GenNCCLIdOp : public framework::OperatorBase {
 
     framework::ProgramDesc empty_program;
     framework::Executor executor(dev_ctx.GetPlace());
```
Comment (luotao1) [inline paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc:216]: gen_nccl_id_op.cc  is used only in GPU, thus, it doesn't need to be updated.

File: paddle/fluid/operators/tensorrt/tensorrt_engine_op.h [hunk 8]
```diff
@@ -134,6 +134,9 @@ class TensorRTEngineOp : public framework::OperatorBase {
   void RunNativeImpl(const framework::Scope &scope,
                      const platform::Place &dev_place) const {
     framework::Executor executor(dev_place);
```
Comment (luotao1) [inline paddle/fluid/operators/tensorrt/tensorrt_engine_op.h:136]: tensorrt_engine_op.h is used only in GPU, thus, it doesn't need to be updated.

**Comment by lidanqing-vv** [comment]
Luotao think this is not elegant. Please consider submitting one issue they will change the executor.

**Comment by grygielski** [comment]
@luotao1 I've submitted an Issue about this problem: https://github.com/PaddlePaddle/Paddle/issues/25988

**Comment by lidanqing-vv** [comment]
This PR and issue #25988 Luotao said they need to discuss inside team. We wait some time.

**Comment by lidanqing-vv** [comment]
@grygielski This PR seems have compatible issues. Test resnet does not pass on windows but on Linux it can pass.  We could ask luotao for help for deploying on Windows and see what is the difference between windows and linux. Why linux pass but windows don't

**Comment by grygielski** [comment]
Closing since https://github.com/PaddlePaddle/Paddle/pull/26502 is already merged

**Pull request closed without merging**